File: frontend/src/api/employees.ts
import type {
  CompanyShare, Employee, EmployeeImportResult, EmployeePosition, EmployeePositionInput,
  EmployeeShares, PayType, UserRole, WeekendPayType,
} from '../types/api'
import { apiClient } from './client'

export interface EmployeeListParams {
  department_id?: number
  is_active?: boolean
  search?: string
}

export const listEmployees = (params?: EmployeeListParams) =>
  apiClient.get<Employee[]>('/api/employees', { params }).then((r) => r.data)

export const getEmployee = (id: number) =>
  apiClient.get<Employee>(`/api/employees/${id}`).then((r) => r.data)

export const createEmployee = (data: {
  tab_number?: string | null
  full_name: string
  position?: string | null
  department_id?: number | null
  schedule_id?: number | null
  default_company_id?: number | null
  pay_type?: PayType
  rate?: string | null
  shift_rate?: string | null
  hour_rate?: string | null
  weekend_pay_type?: WeekendPayType
  weekend_coefficient?: string | null
  weekend_fixed_rate?: string | null
  holiday_pay_type?: WeekendPayType
  holiday_coefficient?: string | null
  holiday_fixed_rate?: string | null
  overtime_coefficient?: string | null
  loan_amount?: string | null
  loan_term_months?: number | null
  loan_start_date?: string | null
  is_active?: boolean
  hire_date?: string | null
  dismissal_date?: string | null
  access?: { email: string; role: UserRole; initial_password: string } | null
}) => apiClient.post<Employee>('/api/employees', data).then((r) => r.data)

export const updateEmployee = (id: number, data: Partial<{
  tab_number: string | null
  full_name: string
  position: string | null
  department_id: number | null
  schedule_id: number | null
  default_company_id: number | null
  pay_type: PayType
  rate: string | null
  shift_rate: string | null
  hour_rate: string | null
  weekend_pay_type: WeekendPayType
  weekend_coefficient: string | null
  weekend_fixed_rate: string | null
  holiday_pay_type: WeekendPayType
  holiday_coefficient: string | null
  holiday_fixed_rate: string | null
  overtime_coefficient: string | null
  loan_amount: string | null
  loan_term_months: number | null
  loan_start_date: string | null
  is_active: boolean
  hire_date: string | null
  dismissal_date: string | null
  is_system_admin: boolean
}>) => apiClient.patch<Employee>(`/api/employees/${id}`, data).then((r) => r.data)

export const deleteEmployee = (id: number) =>
  apiClient.delete(`/api/employees/${id}`)

export const grantAccess = (id: number, data: { email: string; role: UserRole; initial_password: string }) =>
  apiClient.post<Employee>(`/api/employees/${id}/access`, data).then((r) => r.data)

export const updateRole = (id: number, data: { role: UserRole }) =>
  apiClient.patch<Employee>(`/api/employees/${id}/access`, data).then((r) => r.data)

export const resetPassword = (id: number) =>
  apiClient.post<{ temp_password: string }>(`/api/employees/${id}/reset-password`).then((r) => r.data)

export const revokeAccess = (id: number) =>
  apiClient.delete(`/api/employees/${id}/access`)

export const dismissEmployee = (id: number, dismissal_date: string) =>
  apiClient.post<Employee>(`/api/employees/${id}/dismiss`, { dismissal_date }).then((r) => r.data)

export const rehireEmployee = (id: number) =>
  apiClient.post<Employee>(`/api/employees/${id}/rehire`).then((r) => r.data)

// ── Импорт из Excel (task_employee_import) ──

export const downloadImportTemplate = () =>
  apiClient
    .get<Blob>('/api/employees/import/template', { responseType: 'blob' })
    .then((r) => r.data)

/** confirm=false — превью со статусами строк; confirm=true — создать валидные.
 *  Подтверждение шлёт тот же файл: сервер валидирует его заново, а не верит превью. */
export const importEmployees = (file: File, confirm = false) => {
  const form = new FormData()
  form.append('file', file)
  return apiClient
    .post<EmployeeImportResult>('/api/employees/import', form, { params: { confirm } })
    .then((r) => r.data)
}

// ── Позиции (рабочие места) сотрудника — task_positions ч.B ──
// Совместитель = несколько позиций. Читать может любой, кто видит карточку;
// править — только admin.

export const listPositions = (employeeId: number) =>
  apiClient.get<EmployeePosition[]>(`/api/employees/${employeeId}/positions`).then((r) => r.data)

export const createPosition = (employeeId: number, data: EmployeePositionInput) =>
  apiClient.post<EmployeePosition>(`/api/employees/${employeeId}/positions`, data).then((r) => r.data)

export const updatePosition = (
  employeeId: number, positionId: number, data: EmployeePositionInput,
) =>
  apiClient
    .patch<EmployeePosition>(`/api/employees/${employeeId}/positions/${positionId}`, data)
    .then((r) => r.data)

/** Переназначить основную позицию — возвращает весь список в новом порядке. */
export const makePositionPrimary = (employeeId: number, positionId: number) =>
  apiClient
    .post<EmployeePosition[]>(`/api/employees/${employeeId}/positions/${positionId}/make-primary`)
    .then((r) => r.data)

/** «deleted» — позиция удалена; «deactivated» — на ней есть часы/начисления. */
export const deletePosition = (employeeId: number, positionId: number) =>
  apiClient
    .delete<{ result: 'deleted' | 'deactivated' }>(`/api/employees/${employeeId}/positions/${positionId}`)
    .then((r) => r.data)

// ── Распределение по компаниям по умолчанию (задача 3.11b) ──
// Проценты задаются РАБОЧЕМУ МЕСТУ: у совместителя каждое разносится отдельно.
// Без position_id — основная позиция (как было до совместительства).
export const getCompanyShares = (id: number, positionId?: number | null) =>
  apiClient
    .get<EmployeeShares>(`/api/employees/${id}/company-shares`, {
      params: positionId != null ? { position_id: positionId } : undefined,
    })
    .then((r) => r.data)

export const setCompanyShares = (id: number, shares: CompanyShare[], positionId?: number | null) =>
  apiClient
    .put<EmployeeShares>(`/api/employees/${id}/company-shares`, {
      shares, position_id: positionId ?? null,
    })
    .then((r) => r.data)

File: backend/app/services/employees.py
"""
Общий конструктор сотрудника из `EmployeeCreate`.

Вынесен из роутера, чтобы импорт из Excel создавал карточки той же логикой,
что и обычное создание (одни правила по взаимоисключающим полям и дефолтам),
а не собирал `Employee` по-своему. Доступ (email/роль/пароль) сюда НЕ входит —
им управляет роутер.
"""
from __future__ import annotations

from decimal import Decimal

from sqlalchemy.orm import Session

from app.models.employees import Employee
from app.models.positions import (
    PAY_TYPE_HOURLY,
    PAY_TYPE_PER_SHIFT,
    PAY_TYPE_SALARY,
)
from app.schemas.employee import EmployeeCreate


def build_employee(payload: EmployeeCreate) -> Employee:
    return Employee(
        tab_number=payload.tab_number,
        full_name=payload.full_name,
        position=payload.position,
        department_id=payload.department_id,
        schedule_id=payload.schedule_id,
        default_company_id=payload.default_company_id,
        pay_type=payload.pay_type,
        # Оклад, ставка за смену и ставка за час взаимоисключающие — чужие поля
        # не сохраняем, иначе в карточке останется мусор от прошлого типа оплаты
        # и расчёт молча возьмёт не ту базу.
        rate=payload.rate if payload.pay_type == PAY_TYPE_SALARY else None,
        shift_rate=payload.shift_rate if payload.pay_type == PAY_TYPE_PER_SHIFT else None,
        hour_rate=payload.hour_rate if payload.pay_type == PAY_TYPE_HOURLY else None,
        weekend_pay_type=payload.weekend_pay_type,
        # default 1.5 для coefficient, чтобы не хранить NULL при старом поведении
        weekend_coefficient=(
            payload.weekend_coefficient
            if payload.weekend_coefficient is not None or payload.weekend_pay_type != "coefficient"
            else Decimal("1.5")
        ),
        weekend_fixed_rate=payload.weekend_fixed_rate,
        holiday_pay_type=payload.holiday_pay_type,
        # default 1.5 для coefficient — как у выходных, точная ставка в карточке
        holiday_coefficient=(
            payload.holiday_coefficient
            if payload.holiday_coefficient is not None or payload.holiday_pay_type != "coefficient"
            else Decimal("1.5")
        ),
        holiday_fixed_rate=payload.holiday_fixed_rate,
        overtime_coefficient=(
            payload.overtime_coefficient
            if payload.overtime_coefficient is not None
            else Decimal("1.5")
        ),
        loan_amount=payload.loan_amount,
        loan_term_months=payload.loan_term_months,
        loan_start_date=payload.loan_start_date,
        is_active=payload.is_active,
        hire_date=payload.hire_date,
        dismissal_date=payload.dismissal_date,
    )


# ── Табельный номер: уникальность ─────────────────────────────────────────────
# `employees.tab_number` уникален в БД. Без проверки заранее занятый номер
# доходил до Postgres, IntegrityError не ловился и пользователь получал 500.
# Нумерация ОДНА на всю систему (общий справочник и вахта), поэтому и проверка
# одна — ей пользуются оба места создания сотрудника.


def normalize_tab_number(value: str | None) -> str | None:
    """Пустой номер — это «номера нет» (NULL), а не значение.

    Пустая строка уникальна так же, как любая другая: два сотрудника с «» упёрлись
    бы в unique, хотя номера нет ни у одного.
    """
    if value is None:
        return None
    value = value.strip()
    return value or None


def tab_number_conflict(
    db: Session, tab_number: str | None, exclude_id: int | None = None,
) -> str | None:
    """Текст ошибки, если номер уже занят другим сотрудником (включая уволенных)."""
    if tab_number is None:
        return None
    query = db.query(Employee).filter(Employee.tab_number == tab_number)
    if exclude_id is not None:
        query = query.filter(Employee.id != exclude_id)
    holder = query.first()
    if holder is None:
        return None
    suffix = "" if holder.is_active else " (уволен)"
    return f"Табельный номер {tab_number} уже занят: {holder.full_name}{suffix}"
